core/src/SYCL/Kokkos_SYCL_Parallel_Scan.hpp

```diff
@@ -0,0 +1,320 @@
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKO_SYCL_PARALLEL_SCAN_HPP
+#define KOKKO_SYCL_PARALLEL_SCAN_HPP
+
+#include <Kokkos_Macros.hpp>
+#include <memory>
+#if defined(KOKKOS_ENABLE_SYCL)
+
+namespace Kokkos {
+namespace Impl {
+
+template <class FunctorType, class... Traits>
+class ParallelScanSYCLBase {
+ public:
+  using Policy = Kokkos::RangePolicy<Traits...>;
+
+ protected:
+  using Member       = typename Policy::member_type;
+  using WorkTag      = typename Policy::work_tag;
+  using WorkRange    = typename Policy::WorkRange;
+  using LaunchBounds = typename Policy::launch_bounds;
+
+  using ValueTraits = Kokkos::Impl::FunctorValueTraits<FunctorType, WorkTag>;
+  using ValueInit   = Kokkos::Impl::FunctorValueInit<FunctorType, WorkTag>;
+  using ValueJoin   = Kokkos::Impl::FunctorValueJoin<FunctorType, WorkTag>;
+  using ValueOps    = Kokkos::Impl::FunctorValueOps<FunctorType, WorkTag>;
+
+ public:
+  using pointer_type   = typename ValueTraits::pointer_type;
+  using value_type     = typename ValueTraits::value_type;
+  using reference_type = typename ValueTraits::reference_type;
+  using functor_type   = FunctorType;
+  using size_type      = Kokkos::Experimental::SYCL::size_type;
+  using index_type     = typename Policy::index_type;
+
+ protected:
+  const FunctorType m_functor;
+  const Policy m_policy;
+  pointer_type m_scratch_space = nullptr;
+
+ private:
+  template <typename Functor>
+  void scan_internal(cl::sycl::queue& q, const Functor& functor,
+                     pointer_type global_mem, std::size_t size) const {
+    // FIXME_SYCL optimize
+    constexpr size_t wgroup_size = 32;
+    auto n_wgroups               = (size + wgroup_size - 1) / wgroup_size;
+
+    auto deleter = [&q](value_type* ptr) { cl::sycl::free(ptr, q); };
+    std::unique_ptr<value_type[], decltype(deleter)> group_results_memory(
```

Comment:
at a minimum put a todo here that this needs to be a permanent buffer in the Kokkos;:SYCL instance.

Reply:
Agreed.

Reply:
See [`080001a` (#3577)](https://github.com/kokkos/kokkos/pull/3577/commits/080001a1919723601dbfcd22234ca9650e35a2be) and https://github.com/kokkos/kokkos/issues/3585.